src/ray/core_worker/common.h

```diff
@@ -48,20 +49,29 @@ class TaskArg {
     return *id_;
   }
 
-  /// Get the value.
-  std::shared_ptr<Buffer> GetValue() const {
+  /// Get the data of the value.
+  std::shared_ptr<Buffer> GetData() const {
     RAY_CHECK(data_ != nullptr) << "This argument isn't passed by value.";
     return data_;
   }
 
+  /// Get the metadata of the value.
+  std::shared_ptr<Buffer> GetMetadata() const {
+    RAY_CHECK(metadata_ != nullptr) << "This argument isn't passed by value.";
+    return metadata_;
+  }
+
  private:
-  TaskArg(const std::shared_ptr<ObjectID> id, const std::shared_ptr<Buffer> data)
-      : id_(id), data_(data) {}
+  TaskArg(const std::shared_ptr<ObjectID> id, const std::shared_ptr<Buffer> data,
+          const std::shared_ptr<Buffer> metadata)
+      : id_(id), data_(data), metadata_(metadata) {}
 
   /// Id of the argument, if passed by reference, otherwise nullptr.
   const std::shared_ptr<ObjectID> id_;
   /// Data of the argument, if passed by value, otherwise nullptr.
   const std::shared_ptr<Buffer> data_;
+  /// Metadata of the argument, if passed by value, otherwise nullptr.
+  const std::shared_ptr<Buffer> metadata_;
```

Comment:
Is it possible to reuse the existing `RayObject` to present the by-value param in the `TaskArg`? That way it should be more clear and more consistent.

Reply:
also for the data_ and metadata_, could we maintain the following invariants to make the code simpler?
- one of them must be non-null;
- if it's not null, then the buffer must be valid, that is, the size of the buffer must be greater than 0. Thus can only reply on the validity of the pointer itself, without need to further check its size.

Reply:
I'm not sure whether forbidding empty buffers is a better choice. Maybe we can add two more functions, `HasData` and `HasMetadata`.

Do you have any concerns about processing empty buffers?

Reply:
Just found that `RayObject` already has the `HasMetadata` function. I think it's better to ensure that `data_` cannot be nullptr yet is allowed to be empty. As for `metadata_`, we don't set any restrictions.

Reply:
Could you explain a bit more on why it's better to ensure data_ cannot be nullptr?

I think for error conditions `data_` could nullptr?

I preferred to not have empty buffers because:
- it's more intuitive and more clear conceptually. When it's not a nullptr, it's always valid.
- in most cases we only have `data_` but not `metadata_`, and we can save a memory allocation/free in this case.
- I'm a bit concerned that having empty buffers might be error-prone. e.g. it's possible that people might only check the validity of the pointer, but don't check if buffer size is empty, then it might cause some bugs.

Reply:
I've changed my mind. I think we should ensure that the same set of rules apply to both `data_` and `metadata_` to avoid confusion. I tried to add checks in the constructor of `RayObject` to avoid empty buffers, but it just made it worse. Tests failed due to the checks and I can't find out the source of empty buffers. And it cost me hours.

I suggest to keep it as is. We can add more checks after we have an agreement on `nullptr` vs. empty buffer.

Reply:
I see. Then how about just keep it for now, and leave a todo that we can revisit this later.

Reply:
Todo added.

Reply:
I've added checks to avoid empty buffers when constructing a `RayObject` instance.